pkg/cmd/ssh/command_test.go

```diff
@@ -337,6 +366,67 @@ var _ = Describe("Command", func() {
 		Expect(err).To(HaveOccurred())
 	})
 
+	It("should keep the bastion alive", func() {
+		streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+		ctxTimeout, cancelTimeout := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancelTimeout()
+
+		ctx, cancel := context.WithCancel(ctxTimeout)
+		defer cancel()
+
+		// setup fakes
+		currentTarget := target.NewTarget(gardenName, testProject.Name, "", testShoot.Name)
+		targetProvider := internalfake.NewFakeTargetProvider(currentTarget)
+		clientProvider := internalfake.NewFakeClientProvider()
+		clientProvider.WithClient(gardenKubeconfigFile, gardenClient)
+
+		// prepare command
+		factory := internalfake.NewFakeFactory(cfg, nil, clientProvider, nil, targetProvider)
+		factory.ContextImpl = ctx
+
+		options := NewOptions(streams)
+		options.KeepBastion = true // we need to assert its annotations later
+
+		cmd := NewCommand(factory, options)
+
+		// simulate an external controller processing the bastion and proving a successful status
+		go waitForBastionThenPatchStatus(ctx, gardenClient, bastionName, *testProject.Spec.Namespace, func(status *operationsv1alpha1.BastionStatus) {
+			status.Ingress = &corev1.LoadBalancerIngress{
+				Hostname: bastionHostname,
+				IP:       bastionIP,
+			}
+			status.Conditions = []gardencorev1alpha1.Condition{{
+				Type:   "BastionReady",
+				Status: gardencorev1alpha1.ConditionTrue,
+				Reason: "Testing",
+			}}
+		})
+
+		// end the test after a couple of seconds (enough seconds for the keep-alive
+		// goroutine to do its thing)
+		keepAliveInterval = 1 * time.Second
+		createSignalChannel = func() <-chan os.Signal {
+			signalChan := make(chan os.Signal, 1)
+
+			go func() {
+				time.Sleep(5 * time.Second)
+				close(signalChan)
+			}()
+
+			return signalChan
+		}
+
+		// let the magic happen
+		Expect(cmd.RunE(cmd, nil)).To(Succeed())
+
+		key := types.NamespacedName{Name: bastionName, Namespace: *testProject.Spec.Namespace}
+		bastion := &operationsv1alpha1.Bastion{}
+
+		Expect(gardenClient.Get(ctx, key, bastion)).To(Succeed())
+		Expect(bastion.Annotations).To(HaveKeyWithValue(v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationKeepalive))
```

Comment:
Use `Eventually` instead of `Expect`, see https://onsi.github.io/gomega/#making-asynchronous-assertions.
You can then call `close(signalChan)` afterwards and only have something like this as `createSignalChannel`
```go
signalChan := make(chan os.Signal, 1)
createSignalChannel = func() <-chan os.Signal {
	return signalChan
}
```

Maybe it also makes sense to shorten the `keepAliveInterval` to some milliseconds so the test runs even quicker

Reply:
I'm not sure using `Eventually` would work here. If nobody closes the signal channel, the `RunE()` will wait and eventually time out, returning an error.

Reply:
ok then some more changes are necessary. Here I tried it out quick and dirty:

```go
		signalChan := make(chan os.Signal, 1)
		createSignalChannel = func() <-chan os.Signal {
			return signalChan
		}

		key := types.NamespacedName{Name: bastionName, Namespace: *testProject.Spec.Namespace}

		go func() {
			defer GinkgoRecover()

			Eventually(func() bool {
				bastion := &operationsv1alpha1.Bastion{}
				if err := gardenClient.Get(ctx, key, bastion); errors.IsNotFound(err) {
					return false
				}

				return bastion.Annotations != nil && bastion.Annotations[v1beta1constants.GardenerOperation] == v1beta1constants.GardenerOperationKeepalive
			}).Should(BeTrue())
			close(signalChan)
		}()

		// let the magic happen
		Expect(cmd.RunE(cmd, nil)).To(Succeed())

		// Double check that the annotation was really set
		bastion := &operationsv1alpha1.Bastion{}
		Expect(gardenClient.Get(ctx, key, bastion)).To(Succeed())
		Expect(bastion.Annotations).To(HaveKeyWithValue(v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationKeepalive))
```

It has the advantage that the test terminates as soon as the annotation is set - of course depending on the (default) interval of `Eventually`.
It also makes it more resilient in case it takes a bit longer than anticipated (assuming the timeout on `Eventually` is high enough.